onnxruntime/test/providers/qnn/gather_nd_op_test.cc

```diff
@@ -0,0 +1,251 @@
+                                 int expected_nodes_in_graph = 1, int opset = 13) {
+  ProviderOptions provider_options;
+  float fp32_abs_err = 1e-5f;  // default tolerance
+
+#if defined(_WIN32)
+  provider_options["backend_path"] = "QnnHtp.dll";
+#else
+  provider_options["backend_path"] = "libQnnHtp.so";
+#endif
+
+  RunQnnModelTest(BuildQDQGatherNDTestCase<DataType, DataQType>(data_shape, data, indices_shape, indices,
+                                                                indices_are_static, batch_dims),
+                  provider_options,
+                  opset,
+                  expected_ep_assignment,
+                  expected_nodes_in_graph,
+                  test_description,
+                  fp32_abs_err);
+}
+
+//
+// CPU tests:
+//
+
+// Test GatherND op on CPU backend:
+// positive, dynamic indices.
+// QNN EP should support by adding a Cast operator (to int32) after the indices input.
+TEST_F(QnnCPUBackendTests, GatherND_f32_DynamicIndices_BatchDim0) {
+  RunCPUGatherNDOpTest<float>({2, 2, 2},                                        // data_shape
+                              {0.f, 0.1f, 0.2f, 0.3f, 0.4f, 0.5f, 0.6f, 0.7f},  // data
+                              {2, 1, 2},                                        // indices_shape
+                              {0, 1, 1, 0},                                     // indices
+                              false,                                            // indices_are_static
+                              0,                                                // batch_dims
+                              ExpectedEPNodeAssignment::All, "GatherND_f32_DynamicIndices_BatchDim0");
+}
+
+// Test GatherND op on CPU backend:
+// positive, static indices.
+// QNN EP should support by converting static weights to int32_t.
+TEST_F(QnnCPUBackendTests, GatherND_f32_StaticIndices_BatchDim0) {
+  RunCPUGatherNDOpTest<float>({2, 2, 2},                                        // data_shape
+                              {0.f, 0.1f, 0.2f, 0.3f, 0.4f, 0.5f, 0.6f, 0.7f},  // data
+                              {2, 1, 2},                                        // indices_shape
+                              {0, 1, 1, 0},                                     // indices
+                              true,                                             // indices_are_static
+                              0,                                                // batch_dims
+                              ExpectedEPNodeAssignment::All, "GatherND_f32_StaticIndices_BatchDim0");
+}
+
+// Test GatherND op on CPU backend:
+// - positive, dynamic indices.
+// - batch_dims = 1
+// QNN EP should support by adding a Cast operator (to int32) after the indices input.
+//
+// TODO: Enable when QNN fixes GatherNd with batch_dims != 0
+// QNN graph fails to finalized.
+TEST_F(QnnCPUBackendTests, DISABLED_GatherND_f32_DynamicIndices_BatchDim1) {
```

Comment:
This unit test demonstrates potential QNN bug: Graph fails to finalize when the batch_dims parameter is non-zero.